weedcoco/index/indexing.py

```diff
@@ -0,0 +1,142 @@
+            # massage for ES
+            for field in variable_to_null_fields:
+                if agcontext.get(field) == "variable":
+                    del agcontext[field]
+
+        for annotation in coco["annotations"]:
+            image = id_lookup["images", annotation["image_id"]]
+            image.setdefault("annotations", []).append(annotation)
+            annotation["category"] = id_lookup["categories", annotation["category_id"]]
+            # todo: add collection, license
+            _flatten(annotation["category"], annotation, "category")
+            # todo: add collection from collection_memberships
+            image["thumbnail"] = str(
+                self.thumbnail_dir
+                / os.path.basename(image["file_name"])[:2]
+                / os.path.basename(image["file_name"])
+            )
+
+        for image in coco["images"]:
+            try:
+                image["resolution"] = image["width"] * image["height"]
+            except KeyError:
+                pass
+            image["agcontext"] = id_lookup["agcontexts", image["agcontext_id"]]
+            image["sortKey"] = hash(
+                image["file_name"]
+            )  # for deterministic random order
+            _flatten(image["agcontext"], image, "agcontext")
+            # todo: add license
+            image["task_type"] = set()
+            for annotation in image["annotations"]:
+                for k in annotation:
+                    image.setdefault(f"annotation__{k}", []).append(annotation[k])
+
+                # determine available task types
+                image["task_type"].add("classification")
+                if "segmentation" in annotation:
+                    image["task_type"].add("segmentation")
+                    image["task_type"].add("bounding box")
+                if "bbox" in annotation:
+                    image["task_type"].add("bounding box")
+            image["task_type"] = sorted(image["task_type"])
+
+        self.coco = coco
+
+    def generate_batches(self):
+
+        coco = self.coco
+        if "images" in coco and len(coco["images"]) > 0:
+            for i in range(ceil(len(coco["images"]) / self.index_batch)):
+                with open(f"{self.temp_index_dir}/index{str(i+1)}.json", "w") as f:
+                    for image in coco["images"][
+                        i
+                        * self.index_batch : min(
+                            len(coco["images"]), (i + 1) * self.index_batch
+                        )
+                    ]:
+                        json.dump({"index": {"_index": "weedid", "_type": "image"}}, f)
+                        f.write("\n")
+                        json.dump(image, f)
+                        f.write("\n")
+
+    def post_index(self):
+
+        files = [
+            file for file in os.listdir(self.temp_index_dir) if file.endswith(".json")
+        ]
+        for item in files:
+            requests.post(
```

Comment:
It's quite risky to ignore the response status code, for both the query in bulk and individual index entries. The Python elasticsearch package could be used to help manage error codes.

Reply:
I think for now when we only use the bulk post to interact with ElasticSearch and there doesn't show explicit shortage of either speed or capabilities, and the package is bit overkilling with extra parameters and uncertainty to care about. That package is more for enabling different kinds of customised search and XPack APIs which we most likely won't be making use of.
The attributes included in response object should be sufficient in terms of our requirements IMO.